Neo4j.Driver/Neo4j.Driver/Internal/Routing/RoutingTableManager.cs

```diff
@@ -291,31 +291,52 @@ await _discovery.DiscoverAsync(conn, database, impersonatedUser, bookmarks)
                             router, database);
                     }
                 }
-                catch (AuthorizationException e)
+                catch (Exception ex)
                 {
-                    _logger?.Warn(e, "Failed to update routing table from server '{0}' for database '{1}'.", router, database);
-                }
-                catch (SecurityException e)
-                {
-                    _logger?.Error(e,
-                        "Failed to update routing table from server '{0}' for database '{1}' because of a security exception.",
-                        router, database);
-                    throw;
-                }
-                catch (ClientException e)
-                {
-                    _logger?.Error(e,
-                        "Failed to update routing table from server '{0}' for database '{1}' because of a client exception.",
-                        router, database);
-                    throw;
-                }
-                catch (Exception e)
-                {
-                    _logger?.Warn(e, "Failed to update routing table from server '{0}' for database '{1}'.", router, database);
+                    var failfast = IsFailFastException(ex);
+                    var logMsg = "Failed to update routing table from server '{0}' for database '{1}'.";
+                    if (ex is Neo4jException ne)
+                    {
+                        logMsg += " Error code: '{2}'";
+                        var code = ne.Code;
+                        if (failfast)
+                        {
+                            _logger?.Error(ex, logMsg, router, database, code);
+                        }
+                        else
+                        {
+                            _logger?.Warn(ex, logMsg, router, database, code);
+                        }
+                    }
+                    else
+                    {
+                        _logger?.Warn(ex, logMsg, router, database);
+                    }
+
+                    if (failfast)
+                    {
+                        throw;
+                    }
                 }
             }
 
             return null;
         }
+
+        private static bool IsFailFastException(Exception ex)
+        {
+            return ex
+                is FatalDiscoveryException
+                or InvalidBookmarkException
+                or InvalidBookmarkMixtureException
+                or ArgumentErrorException
+                or ProtocolException
+                or TypeErrorException
+                or ForbiddenException
+                or AuthenticationException
+                or TokenExpiredException
+                or ForbiddenException
+                or UnknownSecurityException;
```

Comment:
```suggestion
            return ex
                is FatalDiscoveryException // Neo.ClientError.Database.DatabaseNotFound
                or InvalidBookmarkException // Neo.ClientError.Transaction.InvalidBookmark
                or InvalidBookmarkMixtureException // Neo.ClientError.Transaction.InvalidBookmarkMixture
                or ArgumentErrorException // Neo.ClientError.Statement.ArgumentError
                or ProtocolException // Neo.ClientError.Request.Invalid and (special to .NET driver) Neo.ClientError.Request.InvalidFormat
                or TypeErrorException // Neo.ClientError.Statement.TypeError
                or SecurityException // Neo.ClientError.Security.*
                and not AuthorizationException; // except Neo.ClientError.Security.AuthorizationExpired
```

assuming you change `UnknownSecurityException` to inherit from `SecurityException`, I think this might be shorter, more in line with what the other drivers check, and potentially more future proof as adding more `SecurityException`s later down the road won't require to touch this piece of code that will probably be forgotten until then.

Disclaimer: please check again if this is equivalent, I'm not 100% sure on this one.